addons/mail/models/mail_thread.py

```diff
@@ -2008,112 +2022,226 @@ def message_post(self, *,
+        # preliminary value safety check
+        self._raise_for_invalid_fields([kwargs], ('body', 'composition_mode', 'model', 'res_id', 'values'))
+
+        # with a view, render bodies in batch (template is managed by composer)
+        bodies = self.env['mail.render.mixin']._render_template_qweb_view(
+            view_ref_valid,
+            self._name,
+            self.ids,
+            add_context=render_values,
+        ) if view_ref_valid else {}
+
+        # Prepare composer values for creation
+        composer_values = {
+            'composition_mode': 'mass_mail',
+            'message_type': message_type,
+            'subtype_id': kwargs.pop('subtype_id', False) or self.env['ir.model.data']._xmlid_to_res_id('mail.mt_note'),
+        }
+        composer_values.update(**kwargs)
+        composer_ctx = {
+            'default_composition_mode': 'mass_mail',
+            'default_model': self._name,
+            'default_template_id': template_id if not view_ref else False,
+        }
+
+        mails_su = self.env['mail.mail'].sudo()
+        for subset in [self] if template_id else self:
+            composer_ctx['active_ids'] = subset.ids  # still required for mass mail mode
+            if template_id:
+                composer_ctx['default_res_id'] = subset.ids[0] if subset.ids else 0
             else:
-                kwargs['body'] = rendered_template
-                # a bit complicated to handle, to be improved soon in master
-                _mails_as_sudo, _messages_as_sudo = record.message_post_with_template(False, **kwargs)
-                if _messages_as_sudo:
-                    messages_as_sudo += _messages_as_sudo
-
-        return messages_as_sudo
-
-    def message_post_with_view(self, views_or_xmlid, **kwargs):
-        """ Helper method to send a mail / post a message using a view_id """
-        return self._message_compose_with_view(views_or_xmlid, **kwargs)
-
-    def message_post_with_template(self, template_id, email_layout_xmlid=None, auto_commit=False, **kwargs):
-        """ Helper method to send a mail with a template
-            :param template_id : the id of the template to render to create the body of the message
-            :param **kwargs : parameter to create a mail.compose.message woaerd (which inherit from mail.message)
+                composer_values['body'] = bodies[subset.id]
+                composer_ctx['default_res_id'] = subset.id
+
+            composer = self.env['mail.compose.message'].with_context(
+                **composer_ctx
+            ).create(composer_values)
+            # Simulate the onchange (like trigger in form the view) only
+            # when having a template in single-email mode
+            if template_id:
+                update_values = composer._onchange_template_id(
+                    template_id,
+                    composer_values['composition_mode'],
+                    self._name,
+                    self.ids[0] if self.ids else 0
```

Comment:
same here

Reply:
Yep, bad copy-paste-move that can be cleaned :